service/history/queues/rescheduler.go

```diff
@@ -50,4 +56,93 @@ type (
 		// Len returns the total number of task executables waiting to be rescheduled.
 		Len() int
 	}
+
+	rescheduledExecuable struct {
+		executable     Executable
+		rescheduleTime time.Time
+	}
+
+	reschedulerImpl struct {
+		scheduler  Scheduler
+		timeSource clock.TimeSource
+		scope      metrics.Scope
+
+		sync.Mutex
+		pq collection.Queue[rescheduledExecuable]
+	}
 )
+
+func NewRescheduler(
+	scheduler Scheduler,
+	timeSource clock.TimeSource,
+	scope metrics.Scope,
+) *reschedulerImpl {
+	return &reschedulerImpl{
+		scheduler:  scheduler,
+		timeSource: timeSource,
+		scope:      scope,
+		pq: collection.NewPriorityQueue((func(this rescheduledExecuable, that rescheduledExecuable) bool {
+			return this.rescheduleTime.Before(that.rescheduleTime)
+		})),
+	}
+}
+
+func (r *reschedulerImpl) Add(
+	executable Executable,
+	backoff time.Duration,
+) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.pq.Add(rescheduledExecuable{
+		executable:     executable,
+		rescheduleTime: r.timeSource.Now().Add(backoff),
+	})
+}
+
+func (r *reschedulerImpl) Reschedule(
+	targetRescheduleSize int,
+) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.scope.RecordDistribution(metrics.TaskReschedulerPendingTasks, r.pq.Len())
```

Comment:
is this host level? if yes, I think maybe we should emit this as gauge?

Reply:
No, rescheduler is on shard level, so emitting the metric as a distribution.